packages/data-viz/src/core/HeatmapChart/index.tsx

```diff
@@ -52,22 +52,42 @@ const HeatmapChart = forwardRef(
       throw Error("Chart must have width and height > 0");
     }
 
-    // State for chart initialization
-    const [isChartInitialized, setIsChartInitialized] = useState(false);
-
-    // State for ECharts instance
-    const [chart, setChart] = useState<ECharts | null>(null);
-
     // Ref for the chart container
     const innerRef = useRef<HTMLDivElement | null>(null);
 
+    /**
+     * (thuang): We need both a state and a ref to store the chart instance, so
+     * some hooks can opt out of re-rendering when the chart instance changes.
+     */
+    const [chart, setChart] = useState<ECharts | null>(null);
+    const chartRef = useRef(chart);
+
+    /**
+     * (thuang): Use this ref to store the onEvents prop to prevent
+     * unnecessary re-renders when the onEvents prop changes.
+     * NOTE: This implies that `onEvents` prop changes alone from the parent
```

Comment:
Could this potentially cause issues later on?  
I mean the fact that changing the `onEvents` won't trigger re-renders as you've mentioned!

Reply:
Yeah I was debating about this too 🤔 Really a balance between UX (asking users to `useCallback()` and `useMemo()` to memoize their `onEvents` object, vs. performance (re-render every time `onEvents` changes or not) 

I think in this case it feels like the event handlers are less likely to change than data being passed, so I feel it's probably an okay tradeoff here

Also I think the current implementation in [this PR](https://github.com/chanzuckerberg/sci-components/pull/737/files/f9c6febf8af2ecb68f2f204ae5b759f011d6d817#r1489896229) also ignores all the prop changes except for `[echartsRendererMode]` as well, so changing `onEvents` also won't re-render the chart!